npm/pkg/controlplane/translation/parseSelector.go

```diff
@@ -238,3 +238,158 @@ func parseSelector(selector *metav1.LabelSelector) (labels []string, vals map[st
 
 	return labels, vals
 }
+
+// labelSelector has parsed matchLabels and MatchExpressions information.
+type labelSelector struct {
+	// include is a flag to indicate whether Op exists or not.
+	include bool
+	settype ipsets.SetType
+	// label is among
+	// 1. matchKey + ":" + matchVal (can be empty string) case
+	// 2. "matchKey" case
+	// or 3. "matchKey + : + multiple matchVals" case.
+	label string
+	// members slice exists only if setType is only NestedLabelOfPod.
+	members []string
+}
+
+// parsedSelectors maintains slice of unique labelSelector.
+type parsedSelectors struct {
+	labelSelectors []labelSelector
+	// To avoid the duplilcate matchLabel among matchLabels and MatchExpression
+	// key of labelSet including "!" if operator is "OpNOtIn" or "OpDoesNotExist".
+	labelSet map[string]struct{}
+}
+
+func newParsedSelectors() parsedSelectors {
+	return parsedSelectors{
+		labelSelectors: []labelSelector{},
+		labelSet:       map[string]struct{}{},
+	}
+}
+
+// addSelector only adds non-duplicated labelSelector.
+// Only nested labels from podSelector has members fields.
+func (ps *parsedSelectors) addSelector(include bool, setType ipsets.SetType, matchLabel string, members ...string) {
+	matchLabelWithOp := matchLabel
+	if !include {
+		matchLabelWithOp = "!" + matchLabel
```

Comment:
Does translation need to provide the name with ! operator ? since the include flag is what capturing this condition. DP is not intelligent enough while converting to iptables rules to see if included is false then it adds the "!" infront. translate policy should be relieved of this burden of maintaining matchLabelwith op, 

But if other bits in translate policy needs this matchlabelwithOp then we will have re-design and remove that dependency and follow the single source of truth i.e. included flag, wdyt ?

Reply:
Discussed. Added comments to clarify logic.